### PR TITLE
Removes uses of HTML injection in report

### PR DESCRIPTION
**Pull request by ebidel**
R: @brendankenny @patrickhulce all

Fixes https://github.com/GoogleChrome/lighthouse/issues/1188.

This PR:
- removes the dangerous uses of `{{{}}}`, but keeps the ones where we control the input (`{{{css}}}` and `{{{script}}}`).
- Adds `marked` as a dependency. This module is lightweight (<30kb) and produces the necessary sanitization we need to escape unwanted HTML that gets loaded in the viewer/report.
- Moves the audits to use markdown snippets rather than HTML snippets. Makes `helpText` and `descriptions` shorter and easier to author.
- An added benefit is that we can now use markdown all over the place (aggregation descriptions, audit descriptions, help text). Just need to use `{{sanitize variableName}}`.

@kaycebasques for the FYI as this changes audit help text from html to using markdown.

### CODE REVIEW

File: lighthouse-core/report/report-generator.js [hunk 1]
```diff
@@ -62,15 +63,6 @@ class ReportGenerator {
       return name.toLowerCase().replace(/\s/, '-');
     });
 
-    // Helper for either show an "✘" or "✔" booleans, or simply returning the
-    // value if it's of any other type.
-    Handlebars.registerHelper('getItemValue', value => {
```
Comment (ebidel) [inline lighthouse-core/report/report-generator.js]: we were no longer using this.

File: lighthouse-core/report/report-generator.js [hunk 2]
```diff
@@ -123,6 +115,38 @@ class ReportGenerator {
       }
       return arg;
     });
+
+    // eslint-disable-next-line no-unused-vars
+    Handlebars.registerHelper('sanitize', function(str, opts) {
+      // const isViewer = opts.data.root.reportContext === 'viewer';
```
Comment (ebidel) [inline lighthouse-core/report/report-generator.js]: kept this to remember how to do one offs for different contexts (possible future stuff)

**Comment by patrickhulce** [comment]
nice this will be so much better! need to talk to youtube guy about the dependency?

**Comment by ebidel** [comment]
This version of marked js is checked into g3 third_party :)

On Wed, Dec 28, 2016, 5:01 PM Patrick Hulce <notifications@github.com>
wrote:

nice this will be so much better! need to talk to youtube guy about the
dependency?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/GoogleChrome/lighthouse/pull/1226#issuecomment-269555190>,
or mute the thread

**Comment by ebidel** [comment]
Landing this so the next release has the security improvements and others can move forward with audit changes without running into rebase issues.

**Comment by paulirish** [comment]
marked appears to add just 27kb to the final bundle size. That seems good to me!

image

👍

**Comment by ebidel** [comment]
Yep, very small. That could potentially come down further if we tree shake some unused features out.

**Comment by paulirish** [comment]
Nice. Got an experimental branch up where we do minifying and and viewing this size report: https://github.com/GoogleChrome/lighthouse/compare/minified-sizereport

**Pull request merged**